server/session_ws.go

```diff
@@ -233,13 +237,13 @@ func (s *sessionWS) processOutgoing() {
 
 func (s *sessionWS) pingNow() bool {
 	s.Lock()
+	defer s.Unlock()
```

Comment:
The lock is handled explicitly in a few situations so we can unlock as early as possible, before the function returns. With this change the function will also deadlock the session when it exits due to an error because it calls `s.cleanupClosedConnection()` which tries to acquire its own lock.

I suggest we do not change the locking strategy here to avoid a cascade of further changes needed to make it work.

Reply:
I agree. I made this change because pingTimer.Reset is not protected by the lock. If there's a better way to do this, there's no need to change the locking strategy.